ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -511,6 +516,46 @@ let definition_query (state : State.t) uri position merlin_request =
   let result = location_of_merlin_loc uri result in
   Ok (result, state)
 
+(** [switch_impl_intf path] finds a filepath for a file to switch to based on
+    [path]'s extension or a list of possible filenames to create
+
+    @return a list of single filepath to switch to if it already exists or a
+    list of filepaths to create a file *)
+let switch_impl_intf (state : State.t) (filepath : string) :
+    (Json.t, Jsonrpc.Response.Error.t) result =
+  let uri = Uri.of_path filepath in
+  let open Result.O in
+  let+ doc = Document_store.get state.store uri in
+  let ml, mli, re, rei, mll, mly = ("ml", "mli", "re", "rei", "mll", "mly") in
+  let path_without_extension = Filename.remove_extension filepath ^ "." in
+  let find_switch (exts : string list) =
+    let rec loop acc = function
+      | [] -> acc
+      | ext :: rest ->
+        let file_to_switch_to = path_without_extension ^ ext in
+        if Sys.file_exists file_to_switch_to then
+          [ `String file_to_switch_to ]
+        else
+          loop (`String file_to_switch_to :: acc) rest
+    in
+    `List (loop [] exts |> List.rev)
+  in
+  let extensions_to_switch_to =
+    (* ordering of extensions to switch goes from most to least likely candidate *)
+    match Document.syntax doc with
+    | Ocaml -> (
+      match Document.kind doc with
+      | Intf -> [ ml; re; mly; mll ]
+      | Impl -> [ mli; mll; mly; rei ] )
```

Comment:
`rei`, `mly`, and `mll`.

Reply:
Why is it that we remove re from the list for ocaml intf files but keep rei for ocaml impl files?

Reply:
Just an accidental omission by me. Cross language intf <-> impl jumping is fishy and we shouldn't suggest it to the user. However, I'd like to propose the following:

* If only `foo.ml` exists, jumping to intf will go to `foo.mli`
* If both `foo.ml` and `foo.rei` exists, then we'll respect the user's quirky setup and jump between these two.